[SPARK-39489][CORE] Improve event logging JsonProtocol performance by using Jackson instead of Json4s

JoshRosen opened the pull request:
### What changes were proposed in this pull request?

This PR improves the performance of `org.apache.spark.util.JsonProtocol` by replacing all uses of Json4s with uses of Jackson `JsonGenerator`. In my benchmarking so far, I have seen ~2x speedups for reading events and ~3x speedups for writing them.

JsonProtocol is used by `EventLoggingListener` to emit JSON event logs capturing the firehose of SparkListener events. The history server uses `ReplayListenerBus` to parse the event logs and replay the event stream in order to rebuild and display a terminated Spark application's web UI. 

Today, JsonProtocol uses the Json4s library's ASTs when writing and reading JSON. This existing approach was chosen because we require fine-grained control over how the JSON is generated (so the listener events themselves can't be directly passed to an ObjectMapper). JsonProtocol needs to be backwards- and forwards-compatible so that a single history server application can display UIs from multiple versions of Spark: as a result, the code has special logic for handling fields which might be missing in event logs produced from older Spark versions.

Performance profiling revealed that JsonProtocol spends a significant proportion of its time in Json4s AST calls. Much of this overhead comes from temporary Scala collections which are allocated when constructing the ASTs or when extracting fields from them.

This PR aims to improve performance by replacing all direct uses of Json4s with calls to Jackson's `JsonGenerator`, a low-level imperative API for generating JSON. 

As a simple example, let's consider the `SparkListenerApplicationStart` event: this produces a JSON record like:

```json
{"Event":"SparkListenerApplicationStart","App Name":"Spark shell","App ID":"app-20220615090229-0000","Timestamp":1655308948551,"User":"joshrosen"}
```

With Json4s, this is expressed as:

```scala
def applicationStartToJson(applicationStart: SparkListenerApplicationStart): JValue = {
    ("Event" -> SPARK_LISTENER_EVENT_FORMATTED_CLASS_NAMES.applicationStart) ~
    ("App Name" -> applicationStart.appName) ~
    ("App ID" -> applicationStart.appId.map(JString(_)).getOrElse(JNothing)) ~
    ("Timestamp" -> applicationStart.time) ~
    ("User" -> applicationStart.sparkUser) ~
    ("App Attempt ID" -> applicationStart.appAttemptId.map(JString(_)).getOrElse(JNothing)) ~
    ("Driver Logs" -> applicationStart.driverLogs.map(mapToJson).getOrElse(JNothing)) ~
    ("Driver Attributes" -> applicationStart.driverAttributes.map(mapToJson).getOrElse(JNothing))
  }
```

With Jackson JsonGenerator, this is expressed as:

```scala
def applicationStartToJson(
      applicationStart: SparkListenerApplicationStart,
      g: JsonGenerator): Unit = {
    g.writeStartObject()
    g.writeStringField("Event", SPARK_LISTENER_EVENT_FORMATTED_CLASS_NAMES.applicationStart)
    g.writeStringField("App Name", applicationStart.appName)
    applicationStart.appId.foreach(g.writeStringField("App ID", _))
    g.writeNumberField("Timestamp", applicationStart.time)
    g.writeStringField("User", applicationStart.sparkUser)
    applicationStart.appAttemptId.foreach(g.writeStringField("App Attempt ID", _))
    applicationStart.driverLogs.foreach(writeMapField("Driver Logs", _, g))
    applicationStart.driverAttributes.foreach(writeMapField("Driver Attributes", _, g))
    g.writeEndObject()
  }
```

The JsonGenerator code is more verbose but is significantly faster. This example is a relatively straightforward conversion, but there's a bit of trickiness in more complex cases (such as serialization of task metrics or resource profiles). I will leave comments in the GitHub PR review to highlight some of the less straightforward cases. 

This is a big PR, but most of the changes were made in a semi-automated fashion with regex find-replace macros.

### Why are the changes needed?

Improving JsonProtocol performance has several benefits:

- Decreased history server load times, especially for long-running Spark applications.
- Improved EventLoggingListener throughput: this reduces the likelihood that the will fall behind and drop events (which can lead to confusing and inconsistent Spark UIs).
- Reduced resource consumption on the driver, which may improve the throughput of task scheduling.

In addition, this is a stepping-stone towards eventually being able to remove our Json4s dependency:

- Today Spark uses Json4s 3.x and this causes library conflicts for end users who want to upgrade to 4.x; see https://github.com/apache/spark/pull/33630 for one example.
- To _completely_ remove Json4s we'll need to update several other parts of Spark (including code used for ML model serialization); this PR is just a first step towards that goal if we decide to pursue it.
- In this PR, I continue to use Json4s in test code; I think it's fine to keep Json4s as a test-only dependency.

### Does this PR introduce _any_ user-facing change?
No.


### How was this patch tested?

### Correctness

There are two aspects to consider:

- Ability for old HistoryServer versions to read new logs: I performed tests using randomly-generated events and confirmed that the old and new implementations produce JSON output that is character-for-character identical: see a [review comment](https://github.com/apache/spark/pull/36885#issuecomment-1167855113) for full details.
- Ability to read event logs produced by old Spark versions: I believe this is covered by backwards-compatibility tests in `JsonProtocolSuite`, `SQLJsonProtocolSuite`, and `HistoryServerSuite`. I looked through the git history of `JsonProtocol` and found a few cases where PRs that modified it did not add compatibility tests: for each of these cases, I added new test cases in `JsonProtocolSuite`.

### Performance

So far, I have only measured performance via laptop microbenchmarks.

To measure event parsing performance, I generated a large event log by running `(1 to 5).foreach(_ => spark.range(1, 100 * 1000, 1, 100 * 1000).foreach(_ => ()))` then used the history server's logs to measure the time taken to load the UI. This PR's improvements cut the loading time from ~34 seconds to ~18 seconds.

To measure event logging overheads, I used YourKit asynchronous CPU profiling to to measure `spark.range(1, 100 * 1000, 1, 100 * 1000).foreach(_ => ())`. For this benchmark, this PR's changes cut total time in `logEvent` (including callees) from ~9.5 seconds to ~2.9 seconds.

JoshRosen committed:
Remove from write path.

JoshRosen committed:
Add sparkEventToJsonString helper method.

JoshRosen committed:
Remove from read path.

JoshRosen committed:
WIP towards fixing SQL suites.

JoshRosen committed:
Keep using Json4s for ResourceInfo because it's (unfortunately) part of the DeveloperApi

JoshRosen committed:
Fix writing of "Scope" field.

JoshRosen committed:
Fix "Task Metrics" field

JoshRosen committed:
Fix executorMetricsFromJson: old Spark versions don't write these metrics.

JoshRosen committed:
roll back unnecessary whitespace change

JoshRosen commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -360,255 +460,342 @@ private[spark] object JsonProtocol {
+    g.writeObjectFieldStart("Resources")
+    executorInfo.resourcesInfo.foreach { case (k, v) =>
+      g.writeFieldName(k)
+      g.writeRawValue(compact(v.toJson()))
```
Here we are taking a Json4s `JValue` and are converting it into Json, then writing that raw JSON into a field. I kept this one use of Json4s because I was wary of changing `ResourceInformation.toJson`. That class [is public and is marked as `@Evolving`](https://github.com/apache/spark/blob/9612db3fc9c38204b2bf9f724dedb9ec5f636556/core/src/main/scala/org/apache/spark/resource/ResourceInformation.scala#L28-L40).

In the long run, I don't think that exposing third-party library symbols in our public APIs is a good idea. Perhaps we should make this API private or change it to require that users / implementors return an object which can be turned to JSON via Jackson ObjectMapper.

This particular usage of Json4s isn't critical for performance, so the safest short-term option might be to leave this as-is and consider API changes separately in a followup PR (either targeting the same version if we're okay with changing the API, or deferring to Spark 4.0 when/if we're ready to completely remove our Json4s dependency).

cloud-fan replied:
Shall we leave a TODO comment here?

JoshRosen replied:
Filed https://issues.apache.org/jira/browse/SPARK-39658 as a followup and added code comments referencing it both here in JsonProtocol and on ResourceInformation itself in f324e5e1eb0284ac0ca49c78b5f5c29f8667c131

JoshRosen commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -610,13 +618,34 @@ class JsonProtocolSuite extends SparkFunSuite {
+
+  private val mapper = new ObjectMapper()
+
+  private implicit class JsonStringImplicits(json: String) {
```
Json4s is a convenient way to add/remove fields for compatibility tests. These implicits allow us to do this for strings containing JSON.

JoshRosen commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -610,13 +618,34 @@ class JsonProtocolSuite extends SparkFunSuite {
+    }
+  }
+
+  private implicit def toJsonNode(json: String): JsonNode = {
```
This implicit greatly reduces the amount of boilerplate needed in each test. To me, this seems preferable to having to update hundreds of lines of test cases in this file.

JoshRosen commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -916,13 +958,13 @@ private[spark] object JsonProtocolSuite extends Assertions {
     if (expectedJson != actualJson) {
       // scalastyle:off
       // This prints something useful if the JSON strings don't match
-      println(s"=== EXPECTED ===\n${pretty(expectedJson)}\n")
```
I should probably find a way to preserve the pretty formatting in the debug output. Maybe I can configure the mapper used here in the tests?

JoshRosen commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -360,255 +460,342 @@ private[spark] object JsonProtocol {
-  def propertiesToJson(properties: Properties): JValue = {
-    Option(properties).map { p =>
-      mapToJson(p.asScala)
-    }.getOrElse(JNothing)
```
On re-self-review, I just realized that there's a bug here: in the old code, any fields with `JNothing` values would not be serialized in the final JSON, whereas my code below is returning `writeNull()` in this case. This is a mistake. Instead, I should move the "is properties present" logic one level higher so that I can skip the emission of the "Properties" field if properties is null.

I will fix this tomorrow and will look into writing a test since I don't think we have sufficient coverage here.

JoshRosen replied:
Fixed in https://github.com/apache/spark/pull/36885/commits/c5e59b6d14a3b0f649df3c554e97f5ae7312b74c

JoshRosen commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -663,66 +854,69 @@ private[spark] object JsonProtocol {
-    val taskInfo = taskInfoFromJson(json \ "Task Info")
-    val executorMetrics = executorMetricsFromJson(json \ "Task Executor Metrics")
-    val taskMetrics = taskMetricsFromJson(json \ "Task Metrics")
+      jsonOption(json.get("Stage Attempt ID")).map(_.intValue).getOrElse(0)
```
At the start of this work, I did some toy experiments to see if there were potential behavior differences between Jackson and Json4s when it came to handilng unexpected types (e.g. a field which you expect to contain an int but which contains a string that is an int value, or vice-versa). I think that all of the cases were handled in the same way, but I'd like to reconfirm this.

I'll repeat those experiments and will share the results here.

JoshRosen replied:
I did some experiments in a notebook with code similar to this:

```scala
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._
import org.json4s.DefaultFormats

import com.fasterxml.jackson.core.{JsonEncoding, JsonGenerator}
import com.fasterxml.jackson.databind.{DeserializationFeature, JsonNode, ObjectMapper}
import com.fasterxml.jackson.module.scala.DefaultScalaModule

implicit val format = DefaultFormats

private val mapper = new ObjectMapper().registerModule(DefaultScalaModule)
  .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
```

and test cases like

```scala
import scala.util.Try
val json = """{"value": 21474836470}"""
val json4sResult = Try((parse(json) \ "value").extract[Int])
val jacksonResult = Try(mapper.readTree(json).get("value").intValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
```

This uncovered one subtle corner-case: if you have a string field where a number is expected then my new code will silently return the zero value for the numeric type, whereas the old code would have failed with an exception:

```scala
import scala.util.Try
val json = """{"value": "1"}"""
val json4sResult = Try((parse(json) \ "value").extract[Long])
val jacksonResult = Try(mapper.readTree(json).get("value").longValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
// outputs: AssertionError: assertion failed: json4s: Failure(org.json4s.MappingException: Do not know how to convert JString(1) into long), jackson: Success(0)
```

Similarly, `.textValue` will return `null` for non-Strings:

```scala
import scala.util.Try
val json = """{"value":1}"""
val json4sResult = Try((parse(json) \ "value").extract[String])
val jacksonResult = Try(mapper.readTree(json).get("value").textValue)
assert(json4sResult == jacksonResult, s"json4s: ${json4sResult}, jackson: ${jacksonResult}")
// outputs: AssertionError: assertion failed: json4s: Success(1), jackson: Success(null)
```

These behavior differences aren't an issue if we assume that Spark is always parsing well-formed JSON generated by itself,  but it is a potential risk of silent and confusing behaviors in case Spark reads malformed JSON. Given this, I'm going to explore whether I can write helper functions to add type checking so as to match the old behavior.

JoshRosen replied:
Pushed 03555c6f6074a87c64c059eeddf665b9ffc69c9a to add implicit `extract*` methods that perform typechecking. This has maybe a _slight_ performance impact but the new parsing code is still ~2x faster than the old Json4s code in my local tests.

rednaxelafx reviewed (commented):
Thank you so much for working on this, Josh! This is really useful for us.

Can I ask a newbie questions about version compatibility / optional field handling? The existing code is capable of being lenient in some cases and ignoring certain extra/missing fields. Does the translation from Json4s -> Jackson preserve that behavior?

LuciferYang commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -731,183 +925,180 @@ private[spark] object JsonProtocol {
         ((stageId, stageAttemptId) -> executorMetrics)
       }.toMap
-    }
+    }.getOrElse(Map.empty)
```
seems should be `Map.empty[(Int, Int), ExecutorMetrics]`, otherwise compilation will fail with `Scala-2.13`:

```
[error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1061:64: type mismatch;
[error]  found   : scala.collection.Map[_1,org.apache.spark.executor.ExecutorMetrics] where type _1 <: (Int, Int)
[error]  required: scala.collection.Map[(Int, Int),org.apache.spark.executor.ExecutorMetrics]
[error] Note: _1 <: (Int, Int), but trait Map is invariant in type K.
[error] You may wish to investigate a wildcard type such as `_ <: (Int, Int)`. (SLS 3.2.10)
[error]     SparkListenerExecutorMetricsUpdate(execInfo, accumUpdates, executorUpdates)
[error]                                                                ^
[error] one error found
[error] (core / Compile / compileIncremental) Compilation failed
[error] Total time: 136 s (02:16), completed Jun 16, 2022 2:07:49 AM
```

LuciferYang commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -57,298 +57,398 @@ import org.apache.spark.util.Utils.weakIntern
+  def logStartToJson(logStart: SparkListenerLogStart, g: JsonGenerator): Unit = {
+    g.writeStartObject()
+    g.writeStringField("Event", SPARK_LISTENER_EVENT_FORMATTED_CLASS_NAMES.logStart)
+    g.writeStringField("Spark Version", SPARK_VERSION)
```
"Spark Version" always written as `SPARK_VERSION` instead of  `logStart.sparkVersion`, so it seems that parameter `logStart` is useless

dcoliversun replied:
Sorry, don't agree with that. It's better to leave it as it is. Before PR, `logStartToJson` needs to record `SPARK_VERSION`.

JoshRosen replied:
Yep: this is a pre-existing bug/quirk in the old code, too:

```scala
def logStartToJson(logStart: SparkListenerLogStart): JValue = {
    ("Event" -> SPARK_LISTENER_EVENT_FORMATTED_CLASS_NAMES.logStart) ~
    ("Spark Version" -> SPARK_VERSION)
  }
```

Since my goal in this PR is just to improve performance, I chose to preserve the existing (although confusing) beahvior.

dcoliversun commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -57,298 +57,398 @@ import org.apache.spark.util.Utils.weakIntern
+    Option(taskEnd.taskMetrics).foreach { m =>
+      g.writeFieldName("Task Metrics")
+      taskMetricsToJson(m, g)
+    }
```
Correct me if I'm wrong. Before change, if `taskEnd.taskMetrics` is null, it is `Task Metrics` -> `JNothing`. In your PR, when `taskEnd.taskMetrics` is null, `Task Metrics` field is lost. Is it by design?

JoshRosen replied:
This is correct and matches the Json4s behavior. When writing out an object, Json4s omits fields whose values are `JNothing`. Here's an example illustrating this:

```
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.JsonDSL._

val record = ("Task ID" -> 1) ~ ("Task Metrics" -> JNothing)

println(compact(render(record)))
```

outputs

```
{"Task ID":1}
```

dcoliversun commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -57,298 +57,398 @@ import org.apache.spark.util.Utils.weakIntern
+    val generator = mapper.createGenerator(baos, JsonEncoding.UTF8)
+    block(generator)
+    generator.close()
+    new String(baos.toByteArray, StandardCharsets.UTF_8)
```
Better to close `baos` after `baos.toByteArray`

JoshRosen replied:
Per its Javadoc, `ByteArrayOutputStream.close()` has no effect (the implementation is empty) and I can see a few places in the existing Spark codebase where we don't explicitly close these streams. That said, I will restructure this to close the stream (if only to avoid complaints from static analysis tools).

dcoliversun commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -1730,13 +1729,13 @@ abstract class FsHistoryProviderSuite extends SparkFunSuite with Matchers with L
     val metadata = SparkListenerLogStart(org.apache.spark.SPARK_VERSION)
-    val eventJson = JsonProtocol.logStartToJson(metadata)
-    val metadataJson = compact(eventJson) + "\n"
+    val eventJson = JsonProtocol.sparkEventToJsonString(metadata)
```
Don't use `eventJson` because it's String

JoshRosen committed:
Close ByteArrayOuputStream

JoshRosen committed:
Fix Scala 2.13 compilation

JoshRosen committed:
Fix handling of null `properties` field when writing job and stage start events.

JoshRosen commented:
> Can I ask a newbie questions about version compatibility / optional field handling? The existing code is capable of being lenient in some cases and ignoring certain extra/missing fields. Does the translation from Json4s -> Jackson preserve that behavior?

Yes, it should.

There are two different types of compatibility that we need to consider:

- **Compatibility for reading JSON produced by old versions of Spark:** 
  - When we add a new field to an existing event type, we need the read path to treat that field as optional. Both the old and new code achieve this via the `jsonOption` function, which allows us to control behavior when a field is missing at read time. 
  - There are [some existing tests for this](https://github.com/apache/spark/blob/a859dd25019715165ddb0defe3ddfd8e3cba866e/core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala#L239-L241) in `JsonProtocolSuite`. These tests remove fields from the JSON (simulating JSON generated by old versions that don't emit those fields) and tests that events still parse.
  - There are also some older eventlogs which are used by history server tests: https://github.com/apache/spark/tree/master/core/src/test/resources/spark-events
  - As long as we've been consistent about marking new fields with `jsonOption`, then I think that the existing tests + code review to ensure that we've preserved `jsonOption` should give us pretty high confidence in terms of our ability to parse old logs.
- **Writing event JSON which is readable by old HistoryServer versions**:
  - `JsonProtocolSuite` has tests which serialize events and check that Json4s parses the generated JSON into the same tree as an expected JSON response (stored in a string literal); see [`assertJsonStringEquals`](https://github.com/apache/spark/blob/a859dd25019715165ddb0defe3ddfd8e3cba866e/core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala#L918-L929). Right now this PR is replacing that use of Json4s with Jackson, but I will update this to check that the trees are considered equal by _both_ Jackson and Json4s (and will leave a comment in the code explaining why we have this compatibility test).
  - A stronger guarantee would be to check that the Json4s and Jackson versions of `JsonProtocol` produce character-for-character identical output. I am going to experiment with running a test where I copy the existing `Json4s` version of `JsonProtocol` to a class named `OldJsonProtocol` then generate random events (likely via ScalaCheck) and check that the Json output is exactly identical. If this test passes in all cases then we have strong assurance that our new output will be compatible with old history servers. I'm not planning to check this test in because I don't want to have two versions of JsonProtocol that need to be kept in sync when making updates (and I think the risk of future changes introducing breakage _which wouldn't be caught by another test_ is small).

I think (but will confirm) that our Json4s code path is using Jackson under the hood (behind the Json4s interfaces), which gives me further confidence that things should be compatible; if the underlying json parsers were different then we might have to worry about corner-cases in how numbers are parsed, etc., but hopefully they're using the same low-level reader.

I'll post an update here once I've had a chance to run those more extensive compatibility experiments.

mridulm commented:
+CC @shardulm94, @thejdeep

JoshRosen committed:
Backwards-compatibility test for EnvironmentUpdate

JoshRosen committed:
Backwards-compatibility for TaskInfo.partitionId

JoshRosen committed:
Backwards compatibilty: RDDInfo DeterministicLevel

JoshRosen committed:
Backward compatibility: RDDInfo Barrier

JoshRosen committed:
Add missing assertions in existing "InputMetrics backward compatibility" test

JoshRosen committed:
Add missing assertions in existing "OutputMetrics backward compatibility" test

JoshRosen committed:
Strengten backwards compatibility tests for TaskMetrics

JoshRosen committed:
Backwards compatibility for ExecutorInfo

JoshRosen committed:
Backwards compatibility: TaskKilled

JoshRosen committed:
Backwards compatibility for "Remote Bytes Read To Disk" metric

JoshRosen committed:
Variable rename

JoshRosen committed:
Merge remote-tracking branch 'apache/master' into remove-json4s-from-jsonprotocol

LuciferYang commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -518,22 +582,98 @@ class JsonProtocolSuite extends SparkFunSuite {
+    // The "Metrics Properties" field was added in Spark 3.4.0:
+    val expectedEvent: SparkListenerEnvironmentUpdate = {
+      val e = JsonProtocol.environmentUpdateFromJson(environmentUpdateJsonString)
+      e.copy(environmentDetails = e.environmentDetails.updated("Metrics Properties", Seq.empty))
```
Need a compatible way, for example:

```
e.environmentDetails + ("Metrics Properties" -> Seq.empty[(String, String)])
```

due to `scala.collection.Map` no longer has `updated` method in Scala 2.13

LuciferYang commented on core/src/test/scala/org/apache/spark/util/JsonProtocolSuite.scala:
```diff
@@ -518,22 +582,98 @@ class JsonProtocolSuite extends SparkFunSuite {
+      new ExecutorInfo(
+        "Hostee.awesome.com",
+        11,
+        logUrlMap = Map.empty,
```
`Map.empty[String, String].toMap` is required to be compatible with Scala 2.13

JoshRosen committed:
Scala 2.13 fixes

JoshRosen committed:
More 2.13 fixes

JoshRosen commented:
I submitted https://github.com/apache/spark/pull/37027 to fix the [pre-existing JsonProtocol bugs that I found during my testing](https://github.com/apache/spark/pull/36885#issuecomment-1167855113). Once that PR merges I will fix conflicts in this PR.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into remove-json4s-from-jsonprotocol

JoshRosen committed:
Pretty-print strings in JsonProtocolSuite errors.

JoshRosen committed:
Perform type checking when extracting JSON fields.

JoshRosen commented:
I've fixed merge conflicts and pushed another change to add typechecking during JSON node extraction: this doesn't matter when parsing well-formed / expected JSON, but could make a difference if invalid JSON is presented; see https://github.com/apache/spark/pull/36885#discussion_r911493733 for details.

I removed the "WIP" tag; I think this PR should now be ready for a final round of review.

dcoliversun reviewed (approved):
+1. Great work!!

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala:
```diff
@@ -86,18 +84,17 @@ private[spark] class EventLoggingListener(
-    val metadataJson = compact(eventJson)
-    logWriter.writeEvent(metadataJson, flushLogger = true)
+    val eventJson = JsonProtocol.sparkEventToJsonString(metadata)
+    logWriter.writeEvent(eventJson, flushLogger = true)
```
Not related to this PR: does it mean we generate a big JSON string and write it to files? Is it possible to do it in a streaming fashion?

JoshRosen replied:
Yes, the current code needs to materialize each event as a single JSON string. In principle we could do the write in a streaming fashion by keeping a handle to an output stream. Streaming writes could make it slightly trickier to recover from individual event writing failures, though (e.g. try to serialize but bail out and drop an event if the write fails).

cloud-fan reviewed (approved):
great work!

JoshRosen committed:
Add TODO comment around Json4s in ResourceInformation.toJson

JoshRosen commented:
Thanks everyone. I'm merging this to master (Spark 3.4.0).

(The latest test failure is unrelated / a pre-existing flaky test; my most recent commit only added a code comment)

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -360,256 +466,343 @@ private[spark] object JsonProtocol {
+      g.writeStartObject()
+      g.writeStringField("Declaring Class", line.getClassName)
+      g.writeStringField("Method Name", line.getMethodName)
+      g.writeStringField("File Name", line.getFileName)
```
Hi, @JoshRosen and all.
This seems to break Spark History Server.
```
{"Declaring Class":"org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1","Method Name":"columnartorow_nextBatch_0$","File Name":null,"Line Number":-1}
```

```
3/02/10 16:54:46 ERROR ReplayListenerBus: Exception parsing Spark event log: file:/Users/dongjoon/data/history/eventlog_v2_spark-1676069204164-1qq70hioosynfzib9rmi77wbavnao-driver-job/events_1_spark-1676069204164-1qq70hioosynfzib9rmi77wbavnao-driver-job.zstd
java.lang.IllegalArgumentException: requirement failed: Expected string, got NULL
        at scala.Predef$.require(Predef.scala:281)
        at org.apache.spark.util.JsonProtocol$JsonNodeImplicits.extractString(JsonProtocol.scala:1614)
        at org.apache.spark.util.JsonProtocol$.$anonfun$stackTraceFromJson$1(JsonProtocol.scala:1561)
        at scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
        at scala.collection.Iterator.foreach(Iterator.scala:943)
        at scala.collection.Iterator.foreach$(Iterator.scala:943)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
        at scala.collection.generic.Growable.$plus$plus$eq(Growable.scala:62)
        at scala.collection.generic.Growable.$plus$plus$eq$(Growable.scala:53)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:105)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:49)
        at scala.collection.TraversableOnce.to(TraversableOnce.scala:366)
        at scala.collection.TraversableOnce.to$(TraversableOnce.scala:364)
        at scala.collection.AbstractIterator.to(Iterator.scala:1431)
        at scala.collection.TraversableOnce.toBuffer(TraversableOnce.scala:358)
        at scala.collection.TraversableOnce.toBuffer$(TraversableOnce.scala:358)
        at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1431)
        at scala.collection.TraversableOnce.toArray(TraversableOnce.scala:345)
        at scala.collection.TraversableOnce.toArray$(TraversableOnce.scala:339)
        at scala.collection.AbstractIterator.toArray(Iterator.scala:1431)
        at org.apache.spark.util.JsonProtocol$.stackTraceFromJson(JsonProtocol.scala:1564)
        at org.apache.spark.util.JsonProtocol$.taskEndReasonFromJson(JsonProtocol.scala:1361)
        at org.apache.spark.util.JsonProtocol$.taskEndFromJson(JsonProtocol.scala:938)
        at org.apache.spark.util.JsonProtocol$.sparkEventFromJson(JsonProtocol.scala:876)
        at org.apache.spark.util.JsonProtocol$.sparkEventFromJson(JsonProtocol.scala:865)
        at org.apache.spark.scheduler.ReplayListenerBus.replay(ReplayListenerBus.scala:88)
        at org.apache.spark.scheduler.ReplayListenerBus.replay(ReplayListenerBus.scala:59)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$parseAppEventLogs$3(FsHistoryProvider.scala:1140)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$parseAppEventLogs$3$adapted(FsHistoryProvider.scala:1138)
        at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2777)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$parseAppEventLogs$1(FsHistoryProvider.scala:1138)
        at org.apache.spark.deploy.history.FsHistoryProvider.$anonfun$parseAppEventLogs$1$adapted(FsHistoryProvider.scala:1136)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38)
        at org.apache.spark.deploy.history.FsHistoryProvider.parseAppEventLogs(FsHistoryProvider.scala:1136)
        at org.apache.spark.deploy.history.FsHistoryProvider.rebuildAppStore(FsHistoryProvider.scala:1117)
```

Since this is a release blocker, cc @xinrong-meng , too.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:
```diff
@@ -1218,104 +1416,146 @@ private[spark] object JsonProtocol {
+    json.extractElements.map { line =>
+      val declaringClass = line.get("Declaring Class").extractString
+      val methodName = line.get("Method Name").extractString
+      val fileName = line.get("File Name").extractString
```
This change could be the usual suspect.
```scala
- val fileName = (line \ "File Name").extract[String]
+ val fileName = line.get("File Name").extractString
```

dongjoon-hyun commented:
SPARK-42403 is filed as a release blocker for Apache Spark 3.4.0.

dongjoon-hyun commented:
cc @gengliangwang, @HyukjinKwon , @cloud-fan , @mridulm , too.

JoshRosen commented:
@dongjoon-hyun, thanks for the ping. I think that I spot the problem: it looks like the `File Name` field could contain a `null` value and the old code was likely handling JNothing -> `null.asInstanceOf[String]` implicitly.

I think we can fix this by either modifying `extractString` to handle this case or by special-casing this for `FileName`. I'd prefer the former approach because it's probably more robust.

dongjoon-hyun commented:
Ya, actaully, I tried to relax the `requirement` in `extractString`, but it could miss other cases.
So, I'm currently proposing this one.
- https://github.com/apache/spark/pull/39972

dongjoon-hyun commented:
I'll convert it to a normal PR after adding a test case properly.

JoshRosen commented:
I like your approach as a surgical fix.

I'm trying to think through if there are any other cases where we can get `null` values for strings:

- It can't happen for `Properties` because Java `Properties` forbids `null` values.
- Many of the listener event string fields are things like executorId, block IDs, hostnames, etc., none of which can possibly be null at the time that the event logs are written.
- We _might_ get nulls in the `ExceptionFailure` TaskEndReason because an exception can have a `null` message.

Given this, I'm thinking that I'd prefer to fix this by changing `extractString` to check whether the value is textual or null.

JoshRosen commented:
I've submitted my proposed fix at https://github.com/apache/spark/pull/39973

mridulm commented on core/src/test/scala/org/apache/spark/deploy/history/EventLogTestHelper.scala:
Given `JsonProtocol` is `private[spark]`, unfortunately looks like Azure Log Analytics was depending on private implementation details of spark - which are not gauranteed across versions.

msetodev replied:
It was depending on a previous version of Spark that supported the `sparkEventToJson` method. Was this method just purely working with a JSON object? [UnifiedSparkListener.scala](https://github.com/mspnp/spark-monitoring/blob/main/src/spark-listeners/src/main/scala/org/apache/spark/listeners/UnifiedSparkListener.scala) looks like it hasn't been updated to support the removal of the method. Can perhaps someone who is more skilled at Scala than myself please submit a PR to fix this file? Would be super appreciated.

mridulm replied:
This can be best answered by the authors of the library.

dongjoon-hyun replied:
As mentioned by Bridal, `JsonProtocol.sparkEventToJson` is private. I don't think Apache Spark needs any PR for this, @msetodev .
> It was depending on a previous version of Spark that supported the sparkEventToJson method. Was this method just purely working with a JSON object? 

You can file an issue on https://github.com/mspnp/spark-monitoring.
> Can perhaps someone who is more skilled at Scala than myself please submit a PR to fix this file?
